lib/segment/src/types.rs

```diff
@@ -1185,7 +1187,7 @@ pub enum ValueVariants {
 #[derive(Debug, Deserialize, Serialize, JsonSchema, Clone, PartialEq, Eq)]
 #[serde(untagged)]
 pub enum AnyVariants {
-    Keywords(Vec<String>),
+    Keywords(IndexSet<String, FnvBuildHasher>),
     Integers(Vec<IntPayloadType>),
```

Comment:
Why are integers not optimized?

Reply:
My initial benchmark showed that searching integer iteratively were faster than hashing for a reasonable sized list. I assumed this were the case as integer perfectly fit into cache while being very small. 
Doing a new quick bench however showed

Reply:
Is the threshold the same for both strings and integers? 

Another quick idea (for both strings and integers) is to do binary search over a sorted list as a replacement of iterating. I bet this can increase the threshold before which hashing is better. Although, to be fair, this would add extra complexity and can be addressed later in another PR.

Reply:
Maybe we can base this threshold on number of bytes instead.

> Another quick idea (for both strings and integers) is to do binary search over a sorted list as a replacement of iterating. I bet this can increase the threshold before which hashing is better. Although, to be fair, this would add extra complexity and can be addressed later in another PR.

Yes, I've been thinking about this too. But it would make our logic a bit more complex as we likely end up with two different search methods based on some heuristic.

Reply:
I played around with the comparisons between binary search, linear scan, and IndexSet for searching UUIDs. The results favor the current implementation (the one in this PR 🎉).

Before a small threshold, linear scan and binary search perform basically the same. Y-Axis is nanoseconds, X-Axis is list size.
<img width="1003" alt="imagen" src="https://github.com/qdrant/qdrant/assets/62079184/c4986f8c-b17b-40a1-a4ce-61ab54f58df5">

However maybe the threshold (at least for UUIDs) might be better adjusted at size of 30 items.

TLDR: there is a really small optimization that can be made by setting the threshold based on a byte size heuristic, but I don't think it is worth the complexity.

Reply:
I guess this to be very dependent on hardware configuration.

But what about setting a threshold at 512 bytes (32 * 16 (UUID) = 512)? It might scale nicely that way.